File: python/hardware/processor/cpu.py
from datetime import date
from components.component import Component
from components.hardware import Hardware

class CPU(Hardware, Component):
    def __init__(self, serial_number="", installation_date=None, physically_installed=False, clock_speed_ghz=0.0, cores=0, threads=0, architecture="", cache_size=0.0, hyperthreading=False, id="", manufacturer="", model="", price=0.0, socket_type=""):
        Hardware.__init__(self, serial_number, installation_date, physically_installed)
        Component.__init__(self, id, manufacturer, model, price)
        self._clock_speed_ghz = clock_speed_ghz
        self._cores = cores
        self._threads = threads
        self._architecture = architecture
        self._cache_size = cache_size
        self._hyperthreading = hyperthreading
        self._socket_type = socket_type

    # Setters
    def set_clock_speed_ghz(self, clock_speed_ghz):
        self._clock_speed_ghz = clock_speed_ghz

    def set_cores(self, cores):
        self._cores = cores

    def set_threads(self, threads):
        self._threads = threads

    def set_architecture(self, architecture):
        self._architecture = architecture

    def set_cache_size(self, cache_size):
        self._cache_size = cache_size

    def set_hyperthreading(self, hyperthreading):
        self._hyperthreading = hyperthreading

    def set_socket_type(self, socket_type):
        self._socket_type = socket_type

    # Getters
    def get_clock_speed_ghz(self):
        return self._clock_speed_ghz

    def get_cores(self):
        return self._cores

    def get_threads(self):
        return self._threads

    def get_architecture(self):
        return self._architecture

    def get_cache_size(self):
        return self._cache_size

    def is_hyperthreading(self):
        return self._hyperthreading

    def get_socket_type(self):
        return self._socket_type
    
    def get_model(self):
        return super().get_model()

File: cpp/hardware/processor/CPU.cpp
#pragma once
#include <bits/stdc++.h>

#include "../../component/Hardware.cpp"
#include "../../component/Component.cpp"

class CPU : public Hardware, public Component
{
private:
    double clockSpeedGHz;
    int cores;
    int threads;
    std::string architecture;
    double cacheSize;
    bool hyperthreading;
    std::string socketType;

public:
    // Constructor
    CPU() : Hardware(), Component(), clockSpeedGHz(0.0), cores(0), threads(0), architecture(""), cacheSize(0.0), hyperthreading(false), socketType("") {}
    CPU(const std::string &serial, const Date &installDate, bool installed, double clockSpeed, int coreCount, int threadCount, const std::string &arch, double cache, bool hyper, const std::string &id, const std::string &manufacturer, const std::string &model, double price, const std::string &socket)
        : Hardware(serial, installDate, installed), Component(id, manufacturer, model, price), clockSpeedGHz(clockSpeed), cores(coreCount), threads(threadCount), architecture(arch), cacheSize(cache), hyperthreading(hyper), socketType(socket) {}

    // Destructor
    ~CPU() {}

    // Setters
    void setClockSpeedGHz(double clockSpeed)
    {
        clockSpeedGHz = clockSpeed;
    }

    void setCores(int coreCount)
    {
        cores = coreCount;
    }

    void setThreads(int threadCount)
    {
        threads = threadCount;
    }

    void setArchitecture(const std::string &arch)
    {
        architecture = arch;
    }

    void setCacheSize(double cache)
    {
        cacheSize = cache;
    }

    void setHyperthreading(bool hyper)
    {
        hyperthreading = hyper;
    }

    void setSocketType(const std::string &socket)
    {
        socketType = socket;
    }

    // Getters
    double getClockSpeedGHz() const
    {
        return clockSpeedGHz;
    }

    int getCores() const
    {
        return cores;
    }

    int getThreads() const
    {
        return threads;
    }

    std::string getArchitecture() const
    {
        return architecture;
    }

    double getCacheSize() const
    {
        return cacheSize;
    }

    bool isHyperthreading() const
    {
        return hyperthreading;
    }

    std::string getSocketType() const
    {
        return socketType;
    }
};